text_machina/src/extractors/gap_word.py

```diff
@@ -0,0 +1,109 @@
+from math import ceil
+from random import randint
+from typing import Dict, List, Tuple
+
+from datasets import Dataset
+
+from ..config import InputConfig
+from ..types import TaskType
+from .base import Extractor
+from .utils import spacy_pipeline
+
+
+class GapWord(Extractor):
+    """
+    Extractor that fills the prompt template with a boundary of
+    two word spans (left-side and right-side of a sampled word),
+    and with the number of words the LLM has to generate in
+    between the boundary word spans.
+
+    This extractor needs two template placeholders:
+        - {n}: will be filled with the number of words to generate
+               between the boundary words.
+        - {boundaries}: will be filled with the boundary words separated
+                        by the gap token and newlines.
+                        E.g., "words1 ____ words2"
+
+    This extractor allows to pass the following arguments in the
+    `extractor_args` field from the config:
+        - gap_token (str): gap token, e.g., "____"
+        - max_percentage_boundaries (float): max percentage of
+          boundaries to sample from a text. In a text of N words,
+          there will be N-1 possible boundaries of two word spans.
+        - max_word_span (int): max number of words to be generated
+          between the boundary words.
+        - range_boundary_size (List[float, float]): range where
+          to sample the length of the word spans in the boundaries.
+    """
+
+    def __init__(self, input_config: InputConfig, task_type: TaskType):
+        super().__init__(input_config, task_type)
+        self.args = self.input_config.extractor_args.get("gap_word", {})
+        self.workspace = {
+            "positions": [],
+            "human_spans": [],
+            "num_boundaries": [],
+        }
+
+    def prepare_human(self, human_texts: List[str]) -> List[str]:
+        return [
+            "".join(doc_words) for doc_words in self.workspace["human_spans"]
+        ]
+
+    def _format_boundary(self, pair: Tuple[str, str]) -> str:
+        return f"{pair[0]} {self.args['gap_token']} {pair[1]}"
+
+    def _extract(self, dataset: Dataset) -> Dict[str, List[str]]:
+        text_column = self.input_config.dataset_text_column
+        texts = spacy_pipeline(
+            texts=dataset[text_column], language=self.input_config.language
+        )
+        boundaries = []
+        n_generated_words = []
+        for text in texts:
+            words = list([word.text_with_ws for word in text])
```

Comment:
Do we need the `list`?

Reply:
removed. It was kept from a previous version.